src/metabase/lib/metadata/composed_provider.cljc

```diff
@@ -22,6 +27,44 @@
          (m/distinct-by :id))
         metadata-providers))
 
+(mu/defn fetch-bulk-metadata-with-non-bulk-provider :- [:maybe [:sequential :map]]
+  "Call a non-bulk metadata provider repeatedly to fetch multiple metadatas of the same type."
+  [provider      :- ::lib.schema.metadata/metadata-provider
+   metadata-type :- [:enum :metadata/card :metadata/column :metadata/metric :metadata/segment :metadata/table]
+   ids           :- [:maybe
+                     [:or
+                      [:set ::lib.schema.common/positive-int]
+                      [:sequential ::lib.schema.common/positive-int]]]]
+  (let [f (case metadata-type
+            :metadata/card    metadata.protocols/card
+            :metadata/column  metadata.protocols/field
+            :metadata/metric  metadata.protocols/metric
+            :metadata/segment metadata.protocols/segment
+            :metadata/table   metadata.protocols/table)]
+    (into []
+          (keep (fn [id]
+                  (f provider id)))
+          ids)))
+
+(defn- bulk-metadata [providers metadata-type ids]
+  (loop [[provider & more-providers] providers, unfetched-ids (set ids), fetched []]
+    (cond
+      (empty? unfetched-ids)
+      fetched
+
+      (not provider)
+      fetched
+
+      :else
+      (let [newly-fetched     (if (satisfies? metadata.protocols/BulkMetadataProvider provider)
+                                (metadata.protocols/bulk-metadata provider metadata-type unfetched-ids)
+                                (fetch-bulk-metadata-with-non-bulk-provider provider metadata-type unfetched-ids))
+            newly-fetched-ids (into #{} (map :id) newly-fetched)
+            unfetched-ids     (set/difference unfetched-ids newly-fetched-ids)]
+        (recur more-providers
+               unfetched-ids
+               (into fetched newly-fetched))))))
```

Comment:
This stuff used to live in the QP store namespace in a special `qp.store/bulk-metadata` function; I'm working on phasing out the QP Store altogether (we can just use MLv2 queries as the metadata-providerable going forward), so I moved this here so the composed providers we use in tests and elsewhere can implement bulk provider functionality.

At some point we might just want to rework the protocols since probably everything should support the BulkProvider methods, and at that point there's no reason to have separate methods like `card`, `field`, `metric`, etc anymore... but that's a project for another day